openedx/features/assessment/tests/test_helpers.py

```diff
@@ -0,0 +1,247 @@
+        chapter1_sequential1_vertical1_block1 = modulestore().get_item(chapter1_sequential1_vertical1).children[0]
+
+        anonymous_user = AnonymousUserIdFactory.create(user=self.user, course_id=self.source_course.id)
+
+        # create submission item
+        student_item = StudentItemFactory.create(
+            student_id=anonymous_user.anonymous_user_id,
+            course_id=self.source_course.id,
+            item_id=chapter1_sequential1_vertical1_block1
+        )
+        submission = SubmissionFactory.create(
+            student_item=student_item,
+            created_at=FOUR_DAYS_AGO
+        )
+        student = {
+            'anonymous_user_id': anonymous_user.anonymous_user_id
+        }
+        helpers.autoscore_ora(self.source_course.id, unicode(chapter1_sequential1_vertical1_block1), student)
+
+        # Find the associated rubric for that course_id & item_id
+        rubric_dict = helpers.get_rubric_for_course(self.source_course.id, unicode(chapter1_sequential1_vertical1_block1))
+        options_selected, earned, possible = helpers.select_options(rubric_dict)
+
+        assert mock_log_info.called_once
+        assert mock_reset_score.called_with(anonymous_user.anonymous_user_id,
+                                            self.source_course.id.to_deprecated_string(), unicode(chapter1_sequential1_vertical1_block1))
+        assert mock_set_score.called_with(str(submission.uuid), earned, possible)
+
+    @mock.patch('openedx.features.assessment.helpers.log.info')
+    @mock.patch('openedx.features.assessment.helpers.log.warn')
+    @mock.patch('openedx.features.assessment.helpers.reset_score')
+    @mock.patch('openedx.features.assessment.helpers.set_score')
+    def test_autoscore_ora_with_no_submission(self, mock_log_info, mock_log_war, mock_reset_score, mock_set_score):
+        """ Test auto score without any ora submissions """
+
+        chapter1 = self.course_chapters[0].children[0]
+        chapter1_sequential1 = modulestore().get_item(chapter1).children[0]
+        chapter1_sequential1_vertical1 = modulestore().get_item(chapter1_sequential1).children[0]
+        chapter1_sequential1_vertical1_block1 = modulestore().get_item(chapter1_sequential1_vertical1).children[0]
+
+        anonymous_user = AnonymousUserIdFactory.create(user=self.user, course_id=self.source_course.id)
+
+        student = {
+            'anonymous_user_id': anonymous_user.anonymous_user_id
+        }
+        helpers.autoscore_ora(self.source_course.id, unicode(chapter1_sequential1_vertical1_block1), student)
+
+        assert mock_log_info.not_called
+        assert mock_log_war.called_once
+        assert mock_reset_score.not_called
+        assert mock_set_score.not_called
+
+    def test_select_options_with_two_point_rubric(self):
+        """ Validate rubric points calculation with two points rubric """
+
+        expected_options_selected = {'Ideas': 'Fair', 'Content': 'Fair'}
+        two_points_rubric_expected_earned_points = 4
```

Comment:
We can use ddt libray to make one unit test instead of two tests `test_select_options_with_two_point_rubric ` and `test_select_options_with_three_point_rubric`

Please use ddt to load json from files, instant of constants i.e. `THREE_POINT_RUBRIC_DICTIONARY`, `TWO_POINT_RUBRIC_DICTIONARY`. You can convert JSON to dict in unit test

Reply:
using ddt library and json file is a good idea but putting both test cases into a single one would make stuff a bit complext. As you see we have different expected output results for both two and three points rubrics dict.

Reply:
This is a working test (actually 2 tests). The only thing remaining is to update variable name, test name, etc. Regarding using JSON instead of two RUBRIC_DICTIONARY constant, it is up to you, I think it unnecessary now.

```
@data((TWO_POINT_RUBRIC_DICTIONARY, 4, 4), (THREE_POINT_RUBRIC_DICTIONARY, 4, 6))
@unpack
def test_select_options_with_three_point_rubric(self, RUBRIC_DICTIONARY, return_value_1, return_value_2):
    """ Validate rubric points calculation with two points rubric """

    expected_options_selected = {'Ideas': 'Fair', 'Content': 'Fair'}

    three_points_rubric_expected_earned_points = return_value_1
    three_points_rubric_expected_possible_points = return_value_2

    three_point_rubric_options_selected, three_point_rubric_earned_points, three_point_rubric_possible_points = helpers.select_options(
        RUBRIC_DICTIONARY)

    self.assertEqual(expected_options_selected, three_point_rubric_options_selected)
    self.assertEqual(three_point_rubric_earned_points, three_points_rubric_expected_earned_points)
    self.assertEqual(three_point_rubric_possible_points, three_points_rubric_expected_possible_points)
```

Note: you need to add `@ddt` decorator to class and import all these decorators

Reply:
done